opennav_following/src/following_server.cpp

```diff
@@ -0,0 +1,676 @@
+    target_pose = getPoseAtDistance(object_pose, effective_distance);
+
+    // ... and transform the target_pose into base_frame
+    try {
+      tf2_buffer_->transform(
+        target_pose, target_pose, base_frame_, tf2::durationFromSec(transform_tolerance_));
+    } catch (const tf2::TransformException & ex) {
+      RCLCPP_WARN(get_logger(), "Failed to transform target pose: %s", ex.what());
+      return false;
+    }
+
+    // If the object is behind the robot, we reverse the control
+    auto robot_pose = getRobotPoseInFrame(target_pose.header.frame_id);
+    const double dist = std::hypot(
+      robot_pose.pose.position.x - target_pose.pose.position.x,
+      robot_pose.pose.position.y - target_pose.pose.position.y);
+    bool reversing = allow_backward_ &&
+      dist < desired_distance_ && target_pose.pose.position.x < robot_pose.pose.position.x;
+
+    // Compute and publish controls
+    auto command = std::make_unique<geometry_msgs::msg::TwistStamped>();
+    command->header.stamp = now();
+    if (!controller_->computeVelocityCommand(target_pose.pose, command->twist, true, reversing)) {
+      throw opennav_docking_core::FailedToControl("Failed to get control");
+    }
+    vel_publisher_->publish(std::move(command));
+
+    loop_rate.sleep();
+  }
+  return false;
+}
+
+bool FollowingServer::rotateToObject(geometry_msgs::msg::PoseStamped & object_pose)
+{
+  const double dt = 1.0 / controller_frequency_;
+  auto target_pose = object_pose;
+  target_pose.pose.orientation = nav2_util::geometry_utils::orientationAroundZAxis(
+    tf2::getYaw(target_pose.pose.orientation) + M_PI);
+
+  rclcpp::Rate loop_rate(controller_frequency_);
+  auto start = this->now();
+  auto timeout = rclcpp::Duration::from_seconds(rotate_to_object_timeout_);
+
+  while (rclcpp::ok()) {
+    // Update the iteration start time, used for get robot position, transformation and control
+    iteration_start_time_ = this->now();
+
+    publishFollowingFeedback(FollowObject::Feedback::RETRY);
+
+    // Stop if cancelled/preempted
+    if (checkAndWarnIfCancelled<FollowObject>(following_action_server_, "follow_object") ||
+      checkAndWarnIfPreempted<FollowObject>(following_action_server_, "follow_object"))
+    {
+      return false;
+    }
+
+    // If we completed a full rotation and did not find the object, we stop
+    auto robot_pose = getRobotPoseInFrame(object_pose.header.frame_id);
+    auto angular_distance_to_heading = angles::shortest_angular_distance(
+      tf2::getYaw(robot_pose.pose.orientation), tf2::getYaw(target_pose.pose.orientation));
+    if (fabs(angular_distance_to_heading) < angular_tolerance_) {
+      throw opennav_docking_core::FailedToControl("Failed to rotate to object");
+    }
+
+    // Determine if we find the object
+    if (getRefinedPose(object_pose)) {
+      return true;
+    }
+
+    auto current_vel = std::make_unique<geometry_msgs::msg::TwistStamped>();
```

Comment:
It doesn't strike me that this needs to be a unique pointer. This can just be the `Twist`